std/random.d

```diff
@@ -1361,17 +1620,126 @@ how excellent the source of entropy is.
     }
     else
     {
-        import core.thread : Thread, getpid, MonoTime;
-        static bool seeded;
-        static MinstdRand0 rand;
-        if (!seeded)
+        return cast(uint) .unpredictableSeed!ulong;
+    }
+}
+/// ditto
+template unpredictableSeed(UIntType)
+if (isUnsigned!UIntType)
+{
+    static if (!is(Unqual!UIntType == UIntType))
+        alias unpredictableSeed = .unpredictableSeed!(Unqual!UIntType);
+    else static if (is(UIntType == uint))
+        alias unpredictableSeed = .unpredictableSeed;
+    else
+    {
+        /// ditto
+        @property UIntType unpredictableSeed() @nogc nothrow @trusted
         {
-            uint threadID = cast(uint) cast(void*) Thread.getThis();
-            rand.seed((getpid() + threadID) ^ cast(uint) MonoTime.currTime.ticks);
-            seeded = true;
+            version (AnyARC4Random)
+            {
+                static if (UIntType.sizeof <= uint.sizeof)
+                {
+                    return cast(UIntType) arc4random();
+                }
+                else
+                {
+                    UIntType result = void;
+                    arc4random_buf(&result, UIntType.sizeof);
+                    return result;
+                }
+            }
+            else static if (!is(UIntType == ulong))
+            {
+                return cast(UIntType) unpredictableSeed!ulong;
+            }
+            else
+            {
+                /+
+                In the absence of a convenient arc4random-like interface use a
+                shared static SplitMix generator. Described by Guy L. Steele Jr.,
+                Doug Lea, and Christine H. Flood in "Fast Splittable Pseudorandom
+                Number Generators" (2014). As suggested in the paper use David
+                Stafford's superior parameterizations of the MurmurHash3 64-bit
+                finalizer from
+                http://zimbry.blogspot.com/2011/09/better-bit-mixing-improving-on.html.
+                +/
+                import core.atomic : atomicLoad, atomicOp, atomicStore, MemoryOrder;
+                enum bool has64BitAtomicOps = __traits(compiles,
+                {
+                    shared static ulong example;
+                    atomicOp!"+="(example, ulong.max);
+                });
+
+                static if (has64BitAtomicOps)
+                {
+                    shared static ulong _globalSeed;
+                    shared static ulong _globalGamma;
```

Comment:
You mean putting the variables inside the function rather than outside the function? No reason other than making it immediately obvious that they are only used in one place.

Reply:
No, I'm asking about the rationale for them being `shared`.  The existing `unpredictableSeed` used thread-local storage.  What's the motivation for making this different?

Reply:
Initially I was going to keep them thread-local merely because that's how the previous design worked, but then I asked myself "why make them thread-local?" and I couldn't think of a justification. Using a thread-local PRNG is a performance optimization, but the case it's optimizing is one thread calling `unpredictableSeed` many times while other threads are doing the same thing. In most cases it seems like that would be a misuse of the library.

Reply:
> but then I asked myself "why make them thread-local?" and I couldn't think of a justification

Well, there's another way of looking at that.  Thread-local is the default for static and global variables in D, and with good reason.

So, perhaps the question should be: what is the strong, positive reason for _departing_ from the default option?

Reply:
With a global PRNG as a fallback seeding mechanism the options are to use the output of the global seeder directly, or to use the output of the global seeder to initialize a thread-local seeder then use the output of the thread-local seeder. Both designs have the same number of global variables; the second design just adds thread-locals on top.

Reply:
I'm not sure I follow your reasoning here about the same number of global variables.  Can you explain in more detail?

Reply:
The first design was:
system entropy -> global prng -> thread-local prng -> unpredictableSeed

The revised design was:
system entropy -> global prng -> unpredictableSeed